src/SignatureChecker.sol

```diff
@@ -5,74 +5,118 @@ pragma solidity 0.8.21;
 import { IERC1271 } from "./interfaces/IERC1271.sol";
 
 library SignatureChecker {
-    enum RecoverError {
+    enum Error {
         NoError,
         InvalidSignature,
         InvalidSignatureLength,
-        ExpiredSignature,
         InvalidSignatureS,
-        InvalidSignatureV
+        InvalidSignatureV,
+        SignerMismatch
     }
 
-    function isValidSignature(address signer, bytes32 digest, bytes memory signature) internal view returns (bool) {
+    function decodeECDSASignature(bytes memory signature) internal pure returns (uint8 v, bytes32 r, bytes32 s) {
+        // ecrecover takes the signature parameters, and they can be decoded using assembly.
+        /// @solidity memory-safe-assembly
+        assembly {
+            r := mload(add(signature, 0x20))
+            s := mload(add(signature, 0x40))
+            v := byte(0, mload(add(signature, 0x60)))
+        }
+    }
+
+    function isValidSignature(
+        address signer,
+        bytes32 digest,
+        bytes memory signature
+    ) internal view returns (bool isValid) {
         return isValidECDSASignature(signer, digest, signature) || isValidERC1271Signature(signer, digest, signature);
     }
 
     function isValidECDSASignature(
         address signer,
         bytes32 digest,
         bytes memory signature
-    ) internal view returns (bool) {
-        (RecoverError error, address recovered) = recoverSigner(digest, type(uint256).max, signature);
-        return error == RecoverError.NoError && recovered == signer;
+    ) internal pure returns (bool isValid) {
+        return validateECDSASignature(signer, digest, signature) == Error.NoError;
```

Comment:
Why not decode the signature here and only use one `isValidECDSASignature`?
Same for `recoverECDSASigner` and `validateECDSASignature`.